pkg/util/tas/tas_assignment.go

```diff
@@ -0,0 +1,259 @@
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package tas
+
+import (
+	"iter"
+	"slices"
+
+	"k8s.io/utils/ptr"
+
+	kueue "sigs.k8s.io/kueue/apis/kueue/v1beta2"
+)
+
+type TopologyAssignment struct {
+	Levels  []string
+	Domains []TopologyDomainAssignment
+}
+
+type TopologyDomainAssignment struct {
+	Values []string
+	Count  int32
+}
+
+func valueAtIndex(values kueue.TopologyAssignmentSliceLevelValues, idx int) string {
+	if univ := values.Universal; univ != nil {
+		return *univ
+	}
+	ind := values.Individual
+	prefix := ptr.Deref(ind.CommonPrefix, "")
+	suffix := ptr.Deref(ind.CommonSuffix, "")
+	return prefix + ind.Roots[idx] + suffix
+}
+
+func countAtIndex(slice kueue.TopologyAssignmentSlice, idx int) int32 {
+	if univ := slice.PodCounts.Universal; univ != nil {
+		return *univ
+	}
+	return slice.PodCounts.Individual[idx]
+}
+
+func ValuesAtLevel(ta *kueue.TopologyAssignment, levelIdx int) iter.Seq[string] {
+	if ta == nil {
+		return nil
+	}
+	return func(yield func(string) bool) {
+		for _, slice := range ta.Slices {
+			values := slice.ValuesPerLevel[levelIdx]
+			for i := range int(slice.DomainCount) {
+				if !yield(valueAtIndex(values, i)) {
+					return
+				}
+			}
+		}
+	}
+}
+
+func PodCounts(ta *kueue.TopologyAssignment) iter.Seq[int32] {
+	if ta == nil {
+		return nil
+	}
+	return func(yield func(int32) bool) {
+		for _, slice := range ta.Slices {
+			for i := range int(slice.DomainCount) {
+				if !yield(countAtIndex(slice, i)) {
+					return
+				}
+			}
+		}
+	}
+}
+
+func TotalDomainCount(ta *kueue.TopologyAssignment) int {
+	if ta == nil {
+		return 0
+	}
+	res := 0
+	for _, slice := range ta.Slices {
+		res += int(slice.DomainCount)
+	}
+	return res
+}
+
+func InternalSeqFrom(ta *kueue.TopologyAssignment) iter.Seq[TopologyDomainAssignment] {
+	if ta == nil {
+		return nil
+	}
+	return func(yield func(TopologyDomainAssignment) bool) {
+		for _, slice := range ta.Slices {
+			for i := range int(slice.DomainCount) {
+				req := TopologyDomainAssignment{
+					Count:  countAtIndex(slice, i),
+					Values: make([]string, 0, len(ta.Levels)),
+				}
+				for levelIdx := range ta.Levels {
+					req.Values = append(req.Values, valueAtIndex(slice.ValuesPerLevel[levelIdx], i))
+				}
+				if !yield(req) {
+					return
+				}
+			}
+		}
+	}
+}
+
+func InternalFrom(ta *kueue.TopologyAssignment) *TopologyAssignment {
+	if ta == nil {
+		return nil
+	}
+	return &TopologyAssignment{
+		Levels:  ta.Levels,
+		Domains: slices.Collect(InternalSeqFrom(ta)),
+	}
+}
+
+func fillSingleCompactSliceValues(
+	values *kueue.TopologyAssignmentSliceLevelValues,
+	inputProvider func() iter.Seq[string],
+) {
+	var prefix, suffix string
+	var maxLen, minLen, count int
+	start := true
```

Comment:
This is redundant if you have count already. I see the intention is readability, but this check into count feels very simple anyway